subprojects/plugin-use/src/main/java/org/gradle/plugin/use/internal/DefaultPluginRequestApplicator.java

```diff
@@ -85,63 +88,90 @@ public DefaultPluginRequestApplicator(
             return;
         }
 
         PluginArtifactRepositories resolveContext = pluginRepositoriesProvider.createPluginResolveRepositories();
 
         final PluginResolver effectivePluginResolver = wrapInAlreadyInClasspathResolver(classLoaderScope, resolveContext);
-        if (!requests.isEmpty()) {
+        if (!requests.isEmpty() || !autoAppliedPlugins.isEmpty()) {
             addPluginArtifactRepositories(resolveContext, scriptHandler.getRepositories());
         }
-        List<Result> results = resolvePluginRequests(requests, effectivePluginResolver);
 
         final List<Consumer<PluginManagerInternal>> pluginApplyActions = newLinkedList();
         final Map<Result, PluginImplementation<?>> pluginImplsFromOtherLoaders = newLinkedHashMap();
 
+        List<Result> results = resolvePluginRequests(requests, effectivePluginResolver);
         if (!results.isEmpty()) {
-            for (final Result result : results) {
-                applyPlugin(result.request, result.found.getPluginId(), new Runnable() {
-                    @Override
-                    public void run() {
-                        result.found.execute(new PluginResolveContext() {
-                            @Override
-                            public void addLegacy(PluginId pluginId, Object dependencyNotation) {
-                                pluginApplyActions.add(target -> applyLegacyPlugin(target, result, pluginId));
-                                scriptHandler.addScriptClassPathDependency(dependencyNotation);
-                            }
-
-                            @Override
-                            public void add(PluginImplementation<?> plugin) {
-                                pluginApplyActions.add(target -> applyPlugin(target, result, plugin));
-                            }
-
-                            @Override
-                            public void addFromDifferentLoader(PluginImplementation<?> plugin) {
-                                pluginApplyActions.add(target -> applyPlugin(target, result, plugin));
-                                pluginImplsFromOtherLoaders.put(result, plugin);
-                            }
-                        });
-                        String pluginVersion = result.found.getPluginVersion();
-                        if (pluginVersion != null) {
-                            pluginVersionTracker.setPluginVersionAt(
-                                classLoaderScope,
-                                result.found.getPluginId().getId(),
-                                pluginVersion
-                            );
-                        }
-                    }
-                });
+            applyPlugins(scriptHandler, classLoaderScope, pluginApplyActions, pluginImplsFromOtherLoaders, results);
+        }
+
+        exportBuildLogicClassPathTo(classLoaderScope, scriptHandler.getInstrumentedScriptClassPath());
```

Comment:
Can we lock the classloader scope already here? Then we'd only resolve it in a locked state for checking whether the auto-applied plugins are there and then either use it or throw it away.

Reply:
I guess we can, but that means we have to unlock it. We cannot throw away the classloader scope, just the internal state. And this would work only for the `DefaultClassLoaderScope`